static/src/javascripts/es6/projects/common/modules/crosswords/main.js

```diff
@@ -317,36 +324,56 @@ class Crossword extends React.Component {
         const cells = helpers.cellsForEntry(entry);
 
         if (entry.solution) {
-            _.forEach(cells, (cell, n) => {
-                this.state.grid[cell.x][cell.y].value = entry.solution[n];
+            this.setState({
+                grid: helpers.mapGrid(this.state.grid, (cell, x, y) => {
+                    if (_.some(cells, c => c.x === x && c.y === y)) {
+                        const n = entry.direction === 'across' ?
+                            x - entry.position.x :
+                            y - entry.position.y;
+
+                        cell.value = entry.solution[n];
+                    }
+
+                    return cell;
+                })
             });
-
-            this.forceUpdate();
         }
     }
 
     check (entry) {
-        const cells = _.map(helpers.cellsForEntry(entry), (cell) => this.state.grid[cell.x][cell.y]);
+        const cells = helpers.cellsForEntry(entry);
 
         if (entry.solution) {
             const badCells = _.map(_.filter(_.zip(cells, entry.solution), (cellAndSolution) => {
-                const cell = cellAndSolution[0];
+                const coords = cellAndSolution[0];
+                const cell = this.state.grid[coords.x][coords.y];
                 const solution = cellAndSolution[1];
                 return /^[A-Z]$/.test(cell.value) && cell.value !== solution;
-            }), (cellAndSolution) => cellAndSolution[0]);
-
-            _.forEach(badCells, (cell) => {
-                cell.isError = true;
+            }), cellAndSolution => {
+                return cellAndSolution[0];
             });
 
-            this.forceUpdate();
+            this.setState({
+                grid: helpers.mapGrid(this.state.grid, (cell, gridX, gridY) => {
+                    if (_.some(badCells, bad => bad.x === gridX && bad.y === gridY)) {
+                        cell.isError = true;
+                    }
+
+                    return cell;
+                })
+            });
 
             setTimeout(() => {
-                _.forEach(badCells, (cell) => {
-                    cell.isError = false;
-                    cell.value = '';
+                this.setState({
+                    grid: helpers.mapGrid(this.state.grid, (cell, gridX, gridY) => {
+                        if (_.some(badCells, bad => bad.x === gridX && bad.y === gridY)) {
+                            cell.isError = false;
```

Comment:
we're mutating the cells. `mapGrid` gives you a new grid with a distinct reference but the cells are the same objects as before.